src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducerBase.kt

```diff
@@ -8,79 +8,119 @@ package org.rust.cargo.runconfig.test
+        sourceElement.set(testConfig.sourceElement)
+        configuration.name = testConfig.configurationName
+        val cmd = testConfig.cargoCommandLine().mergeWithDefault(configuration)
         configuration.setFromCmd(cmd)
         return true
     }
 
-    fun findConfig(psi: Array<PsiElement>, climbUp: Boolean = true): TestConfig? {
-        if (psi.isEmpty()) return null
-        createMultipleFileConfiguration(psi.mapNotNull { findElement<RsMod>(it, climbUp) }.toTypedArray())
-            ?.let { return it }
-        findElement<RsFunction>(psi[0], climbUp)
-            ?.let { createSingleConfiguration(it) }
-            ?.let { return it }
-        return findElement<RsMod>(psi[0], climbUp)
-            ?.let { createSingleConfiguration(it) }
+    fun findTestConfig(psi: Array<PsiElement>, climbUp: Boolean = true): TestConfig? =
+        testConfigProviders.asSequence() // for laziness
+            .mapNotNull { provide -> provide(psi, climbUp) }
+            .firstOrNull()
+
+    protected fun registerConfigProvider(provider: TestConfigProvider) {
+        testConfigProviders.add(provider)
     }
 
-    private fun findConfig(context: ConfigurationContext): TestConfig? {
+    private fun findTestConfig(context: ConfigurationContext): TestConfig? {
         val elements = LangDataKeys.PSI_ELEMENT_ARRAY.getData(context.dataContext)
             ?: context.location?.psiElement?.let { arrayOf(it) }
-        return elements?.let { findConfig(it) }
+        return elements?.let { findTestConfig(it) }
     }
 
-    private fun createSingleConfiguration(element: RsQualifiedNamedElement): TestConfig? {
-        if (!isSuitable(element)) return null
-        val configPath = element.crateRelativePath.configPath() ?: return null
-        val target = element.containingCargoTarget ?: return null
-        return TestConfig.SingleTestConfig(commandName, configPath, target, element)
+    private inline fun <reified T : RsQualifiedNamedElement> createConfigFor(
+        psi: Array<PsiElement>,
+        climbUp: Boolean
+    ): TestConfig? {
+        val context = psi
+            .mapNotNull { findElement<T>(it, climbUp) }
+            .singleOrNull { isTestModOrFunction(it) && it.containingCargoTarget != null }
+            ?: return null
+        val configPath = context.crateRelativePath.configPath() ?: return null
+        val target = context.containingCargoTarget ?: return null
+        return SingleItemTestConfig(commandName, configPath, target, context)
     }
 
-    private fun createMultipleFileConfiguration(modules: Array<RsMod>): TestConfig? {
-        val modulesWithBenches = modules.filter { isSuitable(it) && it.containingCargoTarget != null }
+    private fun createConfigForMultipleFiles(psi: Array<PsiElement>, climbUp: Boolean): TestConfig? {
+        val context = psi.mapNotNull { findElement<RsMod>(it, climbUp) }
+        val modsWithTests = context.filter { isTestModOrFunction(it) && it.containingCargoTarget != null }
```

Comment:
Hm, looks like here we want to create test config for multiple files if we can create the corresponding config at least for one atom (function or module) element inside of them.
Can we encapsulate the logic if atom element is suitable for test config creation to some objects and reuse this object for complex elements like a list of psi element or a directory?

Reply:
Not sure I understood, but I have moved common part from `createConfigFor, `createConfigForMultipleFiles` and `createConfigForDirectory` into a separate method + class.

Reply:
Seems to have done.